src/napari/utils/task_status.py

```diff
@@ -0,0 +1,146 @@
+import datetime
+import uuid
+from enum import auto
+from typing import Optional
+
+from napari.utils.misc import Callable, StringEnum
+
+
+class Status(StringEnum):
+    PENDING = auto()
+    BUSY = auto()
+    DONE = auto()
+    CANCELLED = auto()
+    FAILED = auto()
+
+
+class TaskStatusItem:
+    def __init__(
+        self,
+        provider: str,
+        status: Status,
+        description: str,
+        cancel_callback: Optional[Callable] = None,
+    ) -> None:
+        self.id: uuid.UUID = uuid.uuid4()
+        self._provider = provider
+        self._timestamp = [self._timestap()]
+        self._status = [status]
+        self._description = [description]
+        self._cancel_callback = cancel_callback
+
+    def _timestap(self) -> str:
+        return datetime.datetime.now().isoformat()
+
+    def __str__(self) -> str:
+        return f'TaskStatusItem: ({self._provider}, {self.id}, {self._timestamp[-1]}, {self._status[-1]}, {self._description[-1]})'
+
+    def update(self, status: Status, description: str) -> None:
+        self._timestamp.append(self._timestap())
+        self._status.append(status)
+        self._description.append(description)
+
+    def cancel(self) -> bool:
+        self.update(Status.CANCELLED, '')
+        if self._cancel_callback is not None:
+            return self._cancel_callback()
+        return False
+
+    def state(self) -> tuple[str, str, Status, str]:
+        return (
+            self._provider,
+            self._timestamp[-1],
+            self._status[-1],
+            self._description[-1],
+        )
+
+
+class TaskStatusManager:
+    """
+    A task status manager, to store status of long running processes/tasks.
+
+    Only one instance is in general available through napari.
+
+    napari methods and plugins can use it to register and update
+    long running tasks.
+    """
+
+    _tasks: dict[uuid.UUID, TaskStatusItem]
+
+    def __init__(self) -> None:
+        self._tasks: dict[uuid.UUID, TaskStatusItem] = {}
```

Comment:
Any concerns about concurrent access here? Would it be better to have a Queue of small dataclasses that bundle the uuid and the status?